### ISSUE TITLE
Config step CMake fail on Windows

### ISSUE THREAD

**Issue by pip010** [copperspice/doxypress#26]
I have build copperspice but when I try to configure doxypress using cmake-gui I get:

> CMake Error at CMakeLists.txt:22 (include):
  include could not find load file:

  

> xxxxxx/copperspice_BUILD/InstallMinGW.cmake

and indeed I do build copperspice with MSVC but then this !?

**Comment by agserm** [comment]
Thanks for your question. Can you provide a few additional details about your build setup?

1) Can you let us know which version of MSVC you are using?
2) Did you do an installation of CopperSpice after the build process?
3) Did you try adding a `CMAKE_PREFIX_PATH` variable which includes the installed version of CopperSpice when configuring DoxyPress? This path will need to be of the form `<install_dir>/cmake/CopperSpice` to find the appropriate CMake files.

**Comment by pip010** [comment]
1. VS2017 ( very latest )
2. No, I've never done installation of a library under windows, always just point to build folder
3. No, I simply fill the CopperSpice_DIR to where I built copperspice, using gui. BTW it finds the config cmake files but they seem to be incomplete. I built CopperSpice with none of the recomended libs !!!

Here is a more complete log where you can follow the chain of events, already
 CopperSpiceLibraryTargets.cmake is missing from my CopperSpice build dir:

> 
> `CMake Error at C:/Users/ppetrov/Documents/copperspice_BUILD/CopperSpiceConfig.cmake:37 (include):
>   include could not find load file:
> 
>     C:/Users/ppetrov/Documents/copperspice_BUILD/CopperSpiceLibraryTargets.cmake
> Call Stack (most recent call first):
>   CMakeLists.txt:18 (find_package)
> 
> 
> CMake Error at C:/Users/ppetrov/Documents/copperspice_BUILD/CopperSpiceConfig.cmake:38 (include):
>   include could not find load file:
> 
>     C:/Users/ppetrov/Documents/copperspice_BUILD/CopperSpiceBinaryTargets.cmake
> Call Stack (most recent call first):
>   CMakeLists.txt:18 (find_package)
> 
> 
> CMake Error at C:/Users/ppetrov/Documents/copperspice_BUILD/CopperSpiceConfig.cmake:41 (include):
>   include could not find load file:
> 
>     C:/Users/ppetrov/Documents/copperspice_BUILD/CopperSpiceMacros.cmake
> Call Stack (most recent call first):
>   CMakeLists.txt:18 (find_package)
> 
> 
> CMake Error at CMakeLists.txt:22 (include):
>   include could not find load file:
> 
>     C:/Users/ppetrov/Documents/copperspice_BUILD/InstallMinGW.cmake
> 
> 
> CMake Error at src/CMakeLists.txt:221 (COPPERSPICE_RESOURCES):
>   Unknown CMake command "COPPERSPICE_RESOURCES".
> `
>

**Comment by Sp3EdeR** [comment]
The referenced CopperSpiceLibraryTargets.cmake and CopperSpiceBinaryTargets.cmake are generated by the "INSTALL" target of the Copperspice build. So at the very least that must be built beforehand.

**Comment by bgeller** [comment]
We are in the process of dropping the Autotools build system for CopperSpice and DoxyPress.  During the first phase we are enhancing the CopperSpice build files to use modern CMake standards. Once this is in place the build files for DoxyPress will be reviewed and updated.

**Comment by Sp3EdeR** [comment]
I was not able to compile copperspice with MSVC, it had multiple issues.

To work around that and try and save some time, I hacked around the CMake/MSBuild build scripts to integrate the prebuilt version, which I then noticed was built with MinGW and did not seem to work happily together with the MSVC build. (It is probably my error that I wasn't able to properly integrate the prebuilt version into CMake, but I saw no easy solutions for generating the CMake package info that was required.)

So I put together an MSYS2 environment to build doxypress with the hacked prebuilt copperspice libraries. I found the following issues during this on the master branch:

1. In `src/CMakeLists.txt`: Under the `DOXYPRESS_SOURCES` variable, the `${CMAKE_CURRENT_SOURCE_DIR}/emoji_entity.cpp` file is not referenced and is causing linker errors.
2. In `src/CMakeLists.txt`: At least in my MSYS2 environment, the `-lLLVMDemangle`, `-lz` and `-lversion` libraries had to be added to the `if(WIN32)` settings to compile.
3. In `doxy_setup.cpp` / `parseCommandLine`:
   The full argument list is sent into the `for (auto item : dashList) {` loop, where the last argument is the json file, which does not exist in the `argMap`. So the `case INVALID:` branch triggered and closed the program even with a valid command line.

It would be very helpful if you could release your build environment for Copperspice and Doxypress. It is probably my error, but I really was not able to create a properly working environment within a short amount of time. I did read both the Copperspice and Doxypress build instructions in the documentation.

**Comment by bgeller** [comment]
> I was not able to compile copperspice with MSVC, it had multiple issues.

Can you let us know what occurred since MSVC is a supported platform and CopperSpice builds in our CI setup. 

You are correct, our prebuilt binary files for Copperspice and Doxypress were created using MinGW and Autotools.  We are in the process of changing the entire build process. Within the next few weeks all of the Autotools build files will be removed and the CMake build files are updated.  

>  I did read both the Copperspice and Doxypress build instructions in the documentation.

Just checking, did you read about building CopperSpice in the "CS Overview" documentation?  On the page about "Building for Windows (MinGW)" we have links to currently supported versions of MSYS and the MinGW.  

For Items 1 and 2, please note we have not upgraded the CMake files in DoxyPress.  This will occur in the next few weeks, hopefully sooner. 

As for Item 3, this was discovered over the weekend and a new commit is being pushed within the hour.  Great catch.

**Comment by Sp3EdeR** [comment]
Hello,
Thank you for your reply. No, I managed to miss the Copperspice MinGW documentation about the environment. I only read the MSVC and then the Doxypress / MSVC and MinGW documentation.

I'll run an MSVC Copperspice build and send you the errors probably by tomorrow. There are also an insane amount of C4291, C4244 warnings on MSVC, which slow the build considerably (I did turn those off).

**Comment by Sp3EdeR** [comment]
I have uploaded a log of my build results (with the above warnings removed, so that it is readable). Searching for "error " finds the build issues. This was made using MSVC 2019 (16.1.6) with v142 toolchain.

[CopperspiceOutput.zip](https://github.com/copperspice/doxypress/files/3728056/CopperspiceOutput.zip)

**Comment by bgeller** [comment]
Here is one item.  In our CI system we tested MSVC 2017 (15.8.1).  Do you have this set to use C++14?  I am wondering if the newer version is more picky. This is great information and we need to resolve these issues in CopperSpice. 

I understand your company is in a bind and it seems like you need a new build of DoxyPress with the current changes as quickly as possible.  This is a bit unusual but we could build a special version for you as a stop gap. Is a Windows version what you need?

**Comment by Sp3EdeR** [comment]
Hello,
Yes, I used the default C++14 mode. I could give it a run through VS2017. We primarily build with Windows currently, although we've got a cross-platform toolchain set up. A special Windows build would be really great, we'd certainly appreciate it! :)

**Comment by vdeurzen** [comment]
Hi @Sp3EdeR, could you tell me which version of OpenSSL you are currently using? I expect you are running into problems due to an incompatible version of OpenSSL.

**Comment by vdeurzen** [comment]
For the error about narrowing from qint32 to quint32, on GCC this is a warning and as was as I can tell this should, normally, be a warning for msvc too. Our automated builds don't give an error here. Could it be there is some configuration you are using that treats warnings as errors?

From the build output I cannot immediately see why CsMultimedia is failing.

**Comment by bgeller** [comment]
Hi @Sp3EdeR, I have taken the liberty to send you a private email.  Hope this helps.

**Comment by Sp3EdeR** [comment]
Hello @vdeurzen,
Most of the narrowing conversion issues have been warnings indeed, which I've disabled (C4244). I looked into C2398 a bit more and it seems odd. Even Microsoft does not seem to have proper documentation for this error and I've had no luck trying to reproduce in godbolt.org with the same compiler options as in Visual Studio... This one may actually be a compiler bug.

As for OpenSSL, apparently CMake found a distribution that was installed on my computer by an automatic package manager unbeknownst to me. If I'm not mistaken while reading the Conda recipe for the package, it is version 1.1.1c OpenSSL.

Thank you for your tips!

**Comment by vdeurzen** [comment]
For OpenSSL you will need to use 1.0.x at this time, work is underway to provide support for newer versions. I had another good look at the conversion error and as far as I can tell it should be a warning. It's certainly code that needs looking at, but for the time being, I suggest you try to solve the issue with compiler flags :-).

**Comment by agserm** [comment]
> 1. In `src/CMakeLists.txt`: Under the `DOXYPRESS_SOURCES` variable, the `${CMAKE_CURRENT_SOURCE_DIR}/emoji_entity.cpp` file is not referenced and is causing linker errors.

Thanks for reporting this. The DoxyPress CMake files are being reworked and enhanced, and this issue has been resolved. New files will be pushed in about a week.

**Comment by agserm** [comment]
Closing since new files have been pushed and this should be fixed. Please open a new ticket if other build issues occur.
